modin/core/dataframe/algebra/binary.py

```diff
@@ -19,11 +19,98 @@
 from .operator import Operator
 
 
+def coerce_int_to_float64(dtype: np.dtype) -> np.dtype:
+    """
+    Coerce dtype to float64 if it is a variant of integer.
+
+    If dtype is integer, function returns float64 datatype.
+    If not, returns the datatype argument itself.
+
+    Parameters
+    ----------
+    dtype : np.dtype
+        NumPy datatype.
+
+    Returns
+    -------
+    dtype : np.dtype
+        Returns float64 for all int datatypes or returns the datatype itself
+        for other types.
+
+    Notes
+    -----
+    Used to precompute datatype in case of division in pandas.
+    """
+    if dtype in np.sctypes["int"] + np.sctypes["uint"]:
+        return np.dtype(np.float64)
+    else:
+        return dtype
+
+
+def compute_dtypes_common_cast(first, second) -> np.dtype:
+    """
+    Precompute data types for binary operations by finding common type between operands.
+
+    Parameters
+    ----------
+    first : PandasQueryCompiler
+        First operand for which the binary operation would be performed later.
+    second : PandasQueryCompiler
+        Second operand for which the binary operation would be performed later.
+
+    Returns
+    -------
+    dtypes
+        The pandas series with precomputed dtypes.
+
+    Notes
+    -----
+    The dtypes of the operands are supposed to be known.
+    """
+    dtypes_first = dict(zip(first.columns, first._modin_frame._dtypes))
+    dtypes_second = dict(zip(second.columns, second._modin_frame._dtypes))
+    columns_first = set(first.columns)
+    columns_second = set(second.columns)
+    common_columns = columns_first.intersection(columns_second)
+    mismatch_columns = columns_first.union(columns_second) - common_columns
+    # If at least one column doesn't match, the result of the non matching column would be nan.
+    nan_dtype = np.dtype(type(np.nan))
+    dtypes = pandas.Series(
+        [
+            pandas.core.dtypes.cast.find_common_type(
+                [
+                    dtypes_first[x],
+                    dtypes_second[x],
+                ]
+            )
+            for x in common_columns
+        ],
+        index=common_columns,
+    )
+    dtypes = pandas.concat(
+        [
+            dtypes,
+            pandas.Series(
+                [nan_dtype] * (len(mismatch_columns)),
+                index=mismatch_columns,
+            ),
+        ]
+    )
+    dtypes = dtypes.sort_index()
```

Comment:
Why do we have to sort index?

Reply:
Result obtained when performing a binary operation has the combined indices sorted

eg

```
import pandas as pd
df1 = pd.DataFrame({"a": [1, 2],"b":[3,4],"d":[3,4]},dtype=np.float128)  # dtypes: a: int64, b: int64
df2 = pd.DataFrame({"b": [1, 2], "a": [2., 3], "c": [4, 5]},dtype=np.float128) # dtypes a: int64, b: int64, c: int64
df=df1*df2
df.dtypes 
```

Results in 

a    float128
b    float128
c     float64
d     float64
dtype: object